Log broadcaster re-architecture and variable number of confirmations

PiotrTrzpil opened the pull request:
**Overview:**
https://www.notion.so/chainlink/Log-Broadcaster-Rearchitecture-3ea9ffe7b8d84e1786aa4f520b355392

further refactoring and changes after what Spook initially had on the `chore-176418637/log-broadcaster-filtering-2 branch`

**Changes so far:**
+ refactoring:
     -> extract eth subscription related code (without many changes) to eth_subscriber.go
     -> extract listener-related code and log sending (without many changes)  to registrations.go
     -> extract the log queue to pool.go

+ logs (either new ones or those that are waiting to be confirmed more) are now being sent out only as a result of a new head arriving from the HeadTracker, not immediately.

+ added NumSubscriptions on ListenerOpts - this is passed to pool, and also used in registrations.go for a final check before sending the log.

+ for each listener, we keep track of min block number it can receive logs for, to avoid double-sends. In case of chain reorgs, the logic might sometimes lead to multiple sends for the same log, so relying on database is then additional protection (log_broadcasts).

Additionally, all logs should be sent correctly in case of missed heads, as for sends we only check the log heights against the head heights, and not hashes.

TODO: remove trace logging after review
and i'll squash some time before merging

PiotrTrzpil commented:
By the way, should we set num_confirmations for certain subscribers to be larger than 1, in this PR?

spooktheducks reviewed (changes requested):
Already looking like great work! I think `registrations.go` could benefit from more comments (/more detail in the design doc), or alternatively, breaking some of the logic out into well-named methods, because it's some of the most important logic in this PR. I'm going to take another look over this, but before that, let's talk through it a little bit so I'm sure that I'm understanding the intent correctly.

PiotrTrzpil commented on core/internal/cltest/cltest.go:
```diff
@@ -1788,7 +1788,7 @@ func SimulateIncomingHeads(t *testing.T, args SimulateIncomingHeadsArgs) (cleanu
 	chTimeout := time.After(args.Timeout)
 
-	chDone := make(chan struct{})
+	chDone = make(chan struct{})
```
it's not an argument but a return value. But yeah this named returns syntax is quite misleading, i got already bitten by it once - will change it

PiotrTrzpil commented on core/internal/cltest/cltest.go:
```diff
@@ -1809,6 +1809,8 @@ func SimulateIncomingHeads(t *testing.T, args SimulateIncomingHeadsArgs) (cleanu
 					ht.OnNewLongestChain(ctx, *heads[current])
 				}
 				if args.EndBlock >= 0 && current == args.EndBlock {
+					defer func() { recover() }()
```
the main idea is that i needed to expose chDone to the tests to be able to await the end of the incoming heads. 
And, in that case sometimes the cleanup might run earlier than chDone is sent to, so there is a panic when sending.

PiotrTrzpil replied:
though perhaps there is a better typical way of doing that?

samsondav replied:
Can we control when cleanup happens instead?

PiotrTrzpil replied:
ah yes, waiting for chDone and then doing the cleanup solved it - no need for recover indeed

PiotrTrzpil commented on core/services/fluxmonitorv2/integrations_test.go:
```diff
@@ -573,6 +573,10 @@ ds1 -> ds1_parse
 	})
 
+	// Logs from log broadcaster are sent only after a next block is received.
+	time.Sleep(2 * time.Second)
```
yeah i did not find a way to avoid this yet (nor lower the sleep time) - perhaps i need to await the next head, but not sure - will see more

samsondav replied:
Can we not subscribe to the log broadcaster in the test and wait for the requisite log to come through?

PiotrTrzpil replied:
the test already does that, but the issue is: the log will not come unless another head is sent, and not too early, not too late:
```
        // The wait time must be sufficient to allow for flux monitor to finish Register process in log broadcaster
	// and then to have log broadcaster backfill logs after the debounceResubscribe period of ~ 1 sec
	time.Sleep(2 * time.Second)
	// Finally, the logs from log broadcaster are sent only after a next block is received.
	fa.backend.Commit()
```

so to avoid the Sleep, we would have to do something like: `app.LogBroadcaster.AwaitNextDebounceAndBackfill()` and block on that - i'll see what can be done.

PiotrTrzpil replied:
ok so i changed that now to:
```
assert.Eventually(t, func() bool {
		return app.LogBroadcaster.TrackedAddressesCount() >= 2
	}, 3*time.Second, 200*time.Millisecond)
```
after adding that new method. should be much more clear what we are waiting for

PiotrTrzpil commented on core/services/log/broadcaster.go:
```diff
@@ -129,11 +135,12 @@ func (b *broadcaster) awaitInitialSubscribers() {
 			b.onRmSubscribers()
 
 		case <-b.DependentAwaiter.AwaitDependents():
+			<-b.headFromStorageAwaiter.AwaitDependents()
```
i actually tried using DependentAwaiter for this (adding +1 to it) but had no luck for some reason

PiotrTrzpil commented on core/services/log/broadcaster.go:
```diff
@@ -32,35 +29,30 @@ type (
-		IsV2Job() bool
+		headFromStorageAwaiter utils.DependentAwaiter
+		chStop                 chan struct{}
+		chStartupDone          chan struct{}
```
yeah it was 'chDone' but i renamed it to 'chStartupDone' for clarity... 😅  
It works exactly as it did though, it was just that it was confusing to me what exactly is being 'done' - i mean, having chStop and chDone at the same time seemed redundant at first.
But yeah if it's already used widely then i'll rename it back.

PiotrTrzpil commented on core/services/log/broadcaster.go:
```diff
@@ -129,11 +135,12 @@ func (b *broadcaster) awaitInitialSubscribers() {
 
 		case <-b.chStop:
-			close(b.chDone)
+			close(b.chStartupDone)
```
hmm logPool is just a data structure, it does not terminate. And for ethSubscriber, i'm just passing the same chStop to it, so its behavior should be exactly as it was before my changes.
And on chStop, they just lead to a return from startResubscribeLoop, so not sure if it's worth to wait for that?

hm yeah looks like i can `defer close(b.chStartupDone)` on the top of this method, instead of inside startResubscribeLoop

PiotrTrzpil commented on core/services/log/broadcaster.go:
```diff
@@ -302,58 +254,35 @@ func (b *broadcaster) process(chRawLogs <-chan types.Log, chErr <-chan error) (s
+		// We only care about the most recent head
+		x := b.newHeads.RetrieveLatestAndClear()
+		if x == nil {
+			// This should never happen
```
hmm i think nil can only happen if there are multiple readers from this mailbox, which is not the case here, right?

PiotrTrzpil commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,238 @@
+		topic := log.Topic()
+		topics[i] = topic
+
+		// is it ok that there can be only one listener per contract address and topic combination?
```
ah outdated comment, removed now. Yes it does/did support multiple as the listener address is the key in the last map: 
`r.registrations[addr][topic][reg.listener]`

PiotrTrzpil commented on core/services/log/pool.go:
```diff
@@ -0,0 +1,41 @@
+	pool.allLogs = append(pool.allLogs, log)
+}
+
+func (pool *logPool) getLogsToSend(head *models.Head, highestNumConfirmations uint64) []models.Log {
```
yup

PiotrTrzpil commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,238 @@
+				logger.Tracef("No chain saved for listener on address (%v), confirmations: %v", update.toUpdate.opts.Contract.Address(), update.toUpdate.opts.NumConfirmations)
+			}
+			if update.toUpdate.lowestAllowedBlockNumber < update.newLowestAllowedBlockNumber {
+				update.toUpdate.lowestAllowedBlockNumber = update.newLowestAllowedBlockNumber
```
in `*updates = append(*updates, listenerInfoUpdate{info, log.BlockNumber + 1})` but i'll make it explicit

PiotrTrzpil commented:
thanks @spooktheducks ! yeah i extracted some logic into more methods in registrations.go and added some more comments

PiotrTrzpil commented:
I changed the logic related to OnConnected:
- there was OnConnected on Listener interface (and e.g. implemented in FluxMonitor) but it was dead code - never called by log broadcaster, so removed it.
- removed returning 'connected' from Registered felt rather redundant and it was used by callers in a seemingly incorrect way
- the logic related to 'connected' atomic bool in FluxMonitor seemed wrong - it was being set as a result of logBroadcaster.Register() call but in that moment it could sometimes happen that log broadcaster was not connected only for a brief time. Instead, now FluxMonitor checks IsConnected on LB on each poll

samsondav commented on core/services/chainlink/application.go:
```diff
@@ -376,6 +377,8 @@ func (app *ChainlinkApplication) Start() error {
 		return err
 	}
 
+	app.LogBroadcaster.SetLatestHeadFromStorage(app.HeadTracker.HighestSeenHead())
```
This is racey. If the head tracker has pulled down a new head between `Start()` and now, this head number will be wrong. 
You need to read from the DB _before_ the head tracker starts.

PiotrTrzpil replied:
ah yes very true!

PiotrTrzpil replied:
Moved this now to run before the services are started

samsondav commented on core/services/log/eth_subscriber.go:
```diff
@@ -0,0 +1,194 @@
+
+		// Backfill from `backfillDepth` blocks ago.  It's up to the subscribers to
+		// filter out logs they've already dealt with.
+		fromBlock := currentHeight - sub.config.BlockBackfillDepth()
```
Weren't we discussing a move to using the latest head instead? or REPLAY_FROM_BLOCK?

PiotrTrzpil replied:
it does use latest head - just below you have:
```
if latestHeadInDb != nil && uint64(latestHeadInDb.Number) > fromBlock {
			logger.Infow("LogBroadcaster: Using the latest stored head a limit of backfill", "blockNumber", latestHeadInDb.Number, "blockHash", latestHeadInDb.Hash)
			// if the latest head stored in DB is newer, we use it instead as the limit of backfill
			fromBlock = uint64(latestHeadInDb.Number)
		}
```
as for env variables, `we agreed to keep BLOCK_BACKFILL_DEPTH and remove REPLAY_FROM_HEAD`, quoting Spook

samsondav replied:
Ah, got it.

samsondav commented on core/services/log/helpers_test.go:
```diff
@@ -11,26 +25,215 @@ import (
+}
+
+func (listener *simpleLogListener) HandleLog(lb log.Broadcast) {
+	logger.Warnf(">>>>>>>>>>>>>>> Listener %v HandleLog for block %v %v received at %v %v", listener.name, lb.RawLog().BlockNumber, lb.RawLog().BlockHash, lb.LatestBlockNumber(), lb.LatestBlockHash())
```
What's up with all the `>` s?

PiotrTrzpil replied:
Just for visibility in tests

samsondav replied:
Ah I didn't see that this was a `_test.go` file

samsondav commented on core/services/log/helpers_test.go:
```diff
@@ -11,26 +25,215 @@ import (
+	}
+}
+
+func (listener *simpleLogListener) HandleLog(lb log.Broadcast) {
```
Whats the reason for making it a pointer receiver?

PiotrTrzpil replied:
i tend to default to pointer receivers, but yes not needed here - removed

samsondav commented on core/services/log/pool.go:
```diff
@@ -0,0 +1,41 @@
+				logsToKeep = append(logsToKeep, log)
+			}
+		}
+		logger.Tracef("latestBlockNum (%v) > highestNumConfirmations (%v) so deleting older logs. Remaining: %v", latestBlockNum, highestNumConfirmations, len(logsToKeep))
```
Personally I prefer structured logging `logger.Tracew` but you can combine both with `fmt.Sprintf`

PiotrTrzpil replied:
yeah, can change to Tracew - i wasn't sure if we are going to keep those traces though

samsondav replied:
I tend to favour more logging over less, especially since this is at Trace level.

samsondav commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,271 @@
+)
+
+// How it works in general:
+// 1. All received logs are kept in a data structure and deleted ONLY after they are too old for all subscribers
```
What happens if the node crashes or is restarted?

samsondav replied:
What determines if a log is "too old"?

PiotrTrzpil replied:
too old == when given log height is lower than (latest height - highest num_confirmations) - added it to the comment

"What happens if the node crashes or is restarted?" - logs will be backfilled from the latest head from DB, for Subscribers that are added before all dependents of LogBroadcaster are done

If a subscriber is added after the LogBroadcaster does the initial backfill, then it's possible/likely that the backfill fill only have depth:1 (from latest head)

Of course, these backfilled logs + any new logs will only be sent after the num_confirmations for given subscriber.

I'll add that to the comment

samsondav commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,271 @@
+//    A) is too young for that listener
+//    B) the corresponding block height is known to be already processed for that listener
+// So in the normal case, each log will be only processed once, and then its corresponding head will be remembered, so it's not double-sent
+// In case of a re-org, the stored lowestAllowedBlockNumber is reset, so the logs from that chain are considered unprocessed at first.
```
Good doc, I'd like to see this expanded

PiotrTrzpil replied:
added some more info - let me know what still seems unclear?

samsondav commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,271 @@
+// 		Each stored log is then checked against every matched listener and is sent unless:
+//    A) is too young for that listener
+//    B) the corresponding block height is known to be already processed for that listener
+// So in the normal case, each log will be only processed once, and then its corresponding head will be remembered, so it's not double-sent
```
s/head/blockhash ?

PiotrTrzpil replied:
well, currently i'm storing the whole head pointer: `lastSeenChain *models.Head`

samsondav commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,283 @@
+
+// How it works in general:
+//
+// 1. Adding and removing listeners updates the highestNumConfirmations - a number tracking what's the current highest NumConfirmations globally
```
What is NumConfirmations? Is it the same as block confirmations? Why is it affected by adding/removing listeners?

samsondav replied:
Can each listener specify its own minimum required number of confirmations?

samsondav replied:
What does this minimum number of confirmations do?

PiotrTrzpil replied:
`Can each listener specify its own minimum required number of confirmations?` - yes, pretty much - that's NumConfirmations - the main purpose of this PR.

This global highestNumConfirmations is just used internally for keeping track when it's safe to delete logs from the array

PiotrTrzpil replied:
`What is NumConfirmations? Is it the same as block confirmations? ` - yes, number of heads that need to be received before a log from an older block can be sent

samsondav commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,283 @@
+		registrations map[common.Address]map[common.Hash]map[Listener]*listenerMetadata // contractAddress => logTopic => Listener
+		decoders      map[common.Address]AbigenContract
+
+		// highest 'num_confirmations' per all listeners, used to know when to delete old logs
```
This is basically `max(listeners.map(l => l.num_confirmations))` right?

PiotrTrzpil replied:
yes, will clarify

samsondav commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,283 @@
+	return
+}
+
+func (r *registrations) removeSubscriber(reg registration) (needsResubscribe bool) {
```
This does not look thread-safe. Perhaps it doesn't matter? If not can we add a comment explaining why

PiotrTrzpil replied:
it is thread-safe - it's being called only from the main event loop, on receiving a message from a mailbox (onRmSubscribers)

samsondav reviewed (approved):
LGTM pending final review from @spooktheducks

PiotrTrzpil commented:
by the way, to not complicate this PR further, i think https://app.clubhouse.io/chainlinklabs/story/6652 (backfill in batches) should be done in a next PR after this is merged

PiotrTrzpil commented:
added the change suggested by Ryan

RyanRHall commented on core/services/log/pool.go:
```diff
@@ -0,0 +1,45 @@
+func (pool *logPool) getLogsToSend(head *models.Head, highestNumConfirmations uint64, finalityDepth uint64) []types.Log {
+	latestBlockNum := uint64(head.Number)
+	logsToReturn := pool.allLogs
+	logsToKeep := make([]types.Log, 0)
```
nitpick, but the "0" in unnecessary, right?

PiotrTrzpil replied:
ah yes indeed

RyanRHall commented on core/services/log/pool.go:
```diff
@@ -0,0 +1,45 @@
+	keptLogsDepth := finalityDepth
+	if highestNumConfirmations > keptLogsDepth {
+		keptLogsDepth = highestNumConfirmations
+	}
```
:+1: :+1: :+1:

RyanRHall commented on core/services/log/pool.go:
```diff
@@ -0,0 +1,45 @@
+		logger.Tracew("LogBroadcaster: latestBlockNum > highestNumConfirmations so deleting older logs", "latestBlockNum", latestBlockNum, "highestNumConfirmations", highestNumConfirmations, "remainingLogsCount", len(logsToKeep))
+		pool.allLogs = logsToKeep
+	}
+	return logsToReturn
```
why do we return all of the logs in the pool? Shouldn't we want only the logs that met their confirmations and weren't removed due to age?

PiotrTrzpil replied:
well, the check for confirmations is later (in registrations), as it depends on the listener. 
As for removing some due to age, yeah could be done here, but i thought it will be simpler to do all those checks in registrations.go

RyanRHall commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,284 @@
+	for _, perAddress := range r.registrations {
+		for _, perTopic := range perAddress {
+			for _, listener := range perTopic {
+				if listener.opts.NumConfirmations > highestNumConfirmations {
```
wowser!

RyanRHall commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,284 @@
+	}
+}
+
+func copyLog(l types.Log) types.Log {
```
nitpick - but this seems better defined on the log type itself

PiotrTrzpil replied:
yup, true

PiotrTrzpil replied:
oh but Log is from geth actually...

PiotrTrzpil commented on core/services/fluxmonitor/flux_monitor_test.go:
```diff
@@ -196,6 +196,7 @@ func TestPollingDeviationChecker_PollIfEligible(t *testing.T) {
 			fetcher := new(mocks.Fetcher)
 			fluxAggregator := new(mocks.FluxAggregator)
 			logBroadcaster := new(logmocks.Broadcaster)
+			logBroadcaster.On("IsConnected").Return(test.connected).Maybe()
```
hmm do you mean, this can sometimes fail, or is not strict enough?
It's still deterministic, right? Just a bit permissive, but yeah i'll see what can be done

PiotrTrzpil commented on core/services/chainlink/application.go:
```diff
@@ -292,6 +293,12 @@ func NewApplication(config *orm.Config, ethClient eth.Client, advisoryLocker pos
 	}
 	app.HeadTracker = services.NewHeadTracker(store, headTrackables)
 
+	head, err := app.HeadTracker.HighestSeenHeadFromDB()
```
hmm do you have in mind some approach? HeadTracker is created as a last service (and with LB as dependency), so i can't pass it to LB directly

RyanRHall replied:
I do think we can change this, but let's save it for a future PR.

samsondav replied:
@RyanRHall the problem with doing that is it implicitly relies on HeadTracker being started afterwards.

If someone were to move it around in future this would not be obvious and could easily be broken resulting in a rare and difficult to debug race condition.

I think its right to leave it here, and we may even want to add a comment explaining why ordering is important.

PiotrTrzpil replied:
done 👍🏼

PiotrTrzpil commented on core/services/fluxmonitor/flux_monitor.go:
```diff
@@ -519,22 +516,6 @@ func (p *PollingDeviationChecker) Stop() {
 	<-p.waitOnStop
 }
 
-func (p *PollingDeviationChecker) OnConnect() {
```
yeah was quite redundant :)

PiotrTrzpil commented on core/services/fluxmonitor/flux_monitor.go:
```diff
@@ -584,32 +565,29 @@ func (p *PollingDeviationChecker) consume() {
-	} else {
-		p.connected.UnSet()
-	}
+	logger.Info("FluxMonitor: Registered with LogBroadcaster.")
```
true, will remove

PiotrTrzpil commented on core/services/fluxmonitor/flux_monitor_test.go:
```diff
@@ -299,6 +297,7 @@ func TestPollingDeviationChecker_PollIfEligible_Creates_JobSpecErr(t *testing.T)
 	fetcher := new(mocks.Fetcher)
 	fluxAggregator := new(mocks.FluxAggregator)
 	logBroadcaster := new(logmocks.Broadcaster)
+	logBroadcaster.On("IsConnected").Return(true).Maybe()
```
changed some of them now to be more strict

RyanRHall reviewed (dismissed):
Solid work @PiotrTrzpil :tada:

PiotrTrzpil commented:
thanks @RyanRHall !

RyanRHall commented on core/services/keeper/registry_synchronizer_core.go:
```diff
@@ -90,8 +90,9 @@ func (rs *RegistrySynchronizer) Start() error {
 				keeper_registry_wrapper.KeeperRegistryUpkeepRegistered{},
 				keeper_registry_wrapper.KeeperRegistryUpkeepPerformed{},
 			},
+			NumConfirmations: 1,
```
I think this needs to be the configurable `KEEPER_MINIMUM_REQUIRED_CONFIRMATIONS` should be part of the config passed in from the keeper delegate function.

PiotrTrzpil replied:
yeah it is already passed as minConfirmations. I was thinking that we would keep the current logic and change this later when we will be removing e.g. keeper head tracking. But yes i can change it now as well

PiotrTrzpil replied:
yeah it is already passed as minConfirmations. I was thinking that we would keep the current logic and change this later when we will be removing e.g. keeper head tracking. But yes i can change it now as well

spooktheducks reviewed (changes requested):
Brilliant solution, @PiotrTrzpil, this offers the perfect set of guarantees to consumers of this service. Just a few tiny changes + questions. I bet we won't have to do much at all to this logic moving forward :+1: 

I'm also excited to see whether this reduces our DB usage significantly.

PiotrTrzpil commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,273 @@
+	wg.Wait()
+}
+
+/**
```
yup

PiotrTrzpil commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,273 @@
+func (r *registrations) sendLogs(logs []types.Log, orm ORM, latestHead *models.Head) {
+	updates := make([]listenerMetadataUpdate, 0)
+	for _, log := range logs {
+		logger.Tracew("LogBroadcaster: Sending a log", "logBlockNumber", log.BlockNumber)
```
hm it's actually debug logging... 
```
	// Zap does not support trace logging just yet
	Default.Debugw("TRACE: "+msg, keysAndValues...)
```

per package logging will be really helpful here

PiotrTrzpil commented on core/services/log/pool.go:
```diff
@@ -0,0 +1,50 @@
+		keptLogsDepth = highestNumConfirmations
+	}
+	// deleting old logs that will never be sent for any listener anymore
+	if latestBlockNum > keptLogsDepth && len(pool.allLogs) > 0 {
```
yes exactly - which i suppose might be slightly possible on some testnets, or very new chains... But the condition is a bit confusing - i changed the code now to just operate on signed ints

PiotrTrzpil commented on core/services/log/broadcaster.go:
```diff
@@ -133,22 +156,32 @@ func (b *broadcaster) awaitInitialSubscribers() {
-func (b *broadcaster) Register(listener Listener, opts ListenerOpts) (connected bool, unsubscribe func()) {
+func (b *broadcaster) Register(listener Listener, opts ListenerOpts) (unsubscribe func()) {
+	if opts.NumConfirmations > uint64(b.config.EthFinalityDepth()) {
+		logger.Errorf("Registering a listener with NumConfirmations (%v) greater than EthFinalityDepth (%v) will not have any effect",
```
oh, very true - removed the error now

PiotrTrzpil commented on core/services/chainlink/application.go:
```diff
@@ -295,6 +296,16 @@ func NewApplication(config *orm.Config, ethClient eth.Client, advisoryLocker pos
+	}
+	logBroadcaster.SetLatestHeadFromStorage(head)
+
+	// Log Broadcaster waits for other services' registrations
```
👍🏼

PiotrTrzpil commented on core/services/head_tracker.go:
```diff
@@ -600,14 +601,18 @@ func (ht *HeadTracker) unsubscribeFromHead() error {
 }
 
+func (ht *HeadTracker) HighestSeenHeadFromDB() (*models.Head, error) {
+	return ht.store.LastHead(context.Background())
```
looks like yes, based on ht.chDone

PiotrTrzpil commented on core/services/keeper/registry_synchronizer_process_logs.go:
```diff
@@ -78,21 +76,19 @@ func (rs *RegistrySynchronizer) handleUpkeepCanceledLogs(head models.Head, done
 		if was {
 			continue
 		}
-		if !head.IsInChain(broadcast.RawLog().BlockHash) {
```
😅

PiotrTrzpil commented on core/services/log/helpers_test.go:
```diff
@@ -51,3 +255,70 @@ func createJob(t *testing.T, store *store.Store) models.JobSpec {
+}
+
+func (mock *mockEth) subscribeCallCount() int32 {
+	return atomic.LoadInt32(&(mock.subscribeCalls))
```
just some leftover from earlier changes (though strange that goland does not highlight it)

PiotrTrzpil commented on core/services/log/helpers_test.go:
```diff
@@ -51,3 +255,70 @@ func createJob(t *testing.T, store *store.Store) models.JobSpec {
+	return atomic.LoadInt32(&(mock.unsubscribeCalls))
+}
+
+func newMockEthClient(chchRawLogs chan chan<- types.Log, blockHeight int64, timesSubscribe int) *mockEth {
```
was an interesting reduction in complexity, yeah :)

PiotrTrzpil commented:
hm i can't respond to https://github.com/smartcontractkit/chainlink/pull/4108#discussion_r617010884 for some reason, so will write here:
```
0 actually means "no limit" (although previously, there was a bug in that code, now fixed). It's only the channel inside the mailbox that ends up buffered, because reasonably, it has to have some limit or else we could end up deadlocking any component that relies on mailboxes.

Happy to consider bumping the fallback channel buffer size, though, or disallowing 0 entirely. The latter is something I've been considering myself.
```

So 0 means the queue is unbounded and has initial capacity 100. 
But what do you mean by "fallback channel buffer size" ?

Anyway, i'll add that to 'Mailbox improvements" (story 7236)

PiotrTrzpil commented on core/services/log/broadcaster.go:
```diff
@@ -120,6 +142,7 @@ func (b *broadcaster) Stop() error {
 }
 
 func (b *broadcaster) awaitInitialSubscribers() {
+	defer close(b.chDone)
```
ah yes, the `go b.startResubscribeLoop()`return  was not waited upon - yup, wgDone is more clear here.

PiotrTrzpil commented on core/services/log/broadcaster.go:
```diff
@@ -88,29 +92,47 @@ var _ Broadcaster = (*broadcaster)(nil)
+
 func (b *broadcaster) Start() error {
 	return b.StartOnce("Log broadcaster", func() error {
+		if b.latestHead != nil {
```
not easily - not without having some weird cross-domain DB access. See discussion there

spooktheducks commented:
> So 0 means the queue is unbounded and has initial capacity 100.
> But what do you mean by "fallback channel buffer size" ?
> 
> Anyway, i'll add that to 'Mailbox improvements" (story 7236)

The mailbox's internal channel is buffered, which is what gives the mailbox its fundamental useful property of never blocking upon delivery of a message. Normally, the channel buffer is the same size as the slice that holds the values, except in the case of 0. In that case, I chose to give the channel some kind of non-zero default, but the value of that default is totally arbitrary. Happy to talk about changing it.

RyanRHall commented on core/services/head_tracker.go:
```diff
@@ -600,14 +601,20 @@ func (ht *HeadTracker) unsubscribeFromHead() error {
 }
 
+func (ht *HeadTracker) HighestSeenHeadFromDB() (*models.Head, error) {
+	ctx, cancel := utils.ContextFromChan(ht.done)
```
this should probably be a combined context that includes `postgres.DefaultQueryCtx()`

PiotrTrzpil replied:
done 👍🏼

RyanRHall commented on core/services/log/registrations.go:
```diff
@@ -0,0 +1,261 @@
+		decoders      map[common.Address]AbigenContract
+
+		// highest 'NumConfirmations' per all listeners, used to decide when it's safe to delete older logs
+		// it's: max(listeners.map(l => l.num_confirmations)
```
this comment could use an update

PiotrTrzpil replied:
updated to mention finality depth
